### ISSUE TITLE
Fuzzy search patterns

### ISSUE THREAD

**Issue by christophrumpel** [mpociot/botman#207]
Hey Marcel,

I use the callback patterns a lot, because it is easy to store my callbacks outside of the conversation and this keeps my conversation class cleaner.

**The Problem**

Callbacks are only matched when the correct `keyword` was sent by the user. If he has a little typo it does not work.

You once said I could use a BotMan middleware, but I don't see how this is working here because I need the `pattern` and the users `message` in order to compare them.

**Solution**

A solution would be to allow `fuzzy patterns`. Maybe something like this: (just from the users view without checking if it would work)

```php
'pattern' => 'joke',
'fuzzy' => 1
'callback' => function() {
        $this->say('Joke it its');
}
```

The fuzzy param could be a value how close the words should be.

Behind the back we would need to check through all the given patterns and the message to see if one of the fits with the fuzzy comparison.

If you're interested in this I could try to find a solution.

Cheers

PS: Of course I could implement fuzzy search myself in my application when I don't use the callback patterns but it would be great to find a way with the patterns.

**Comment by mpociot** [comment]
Hi Christoph,

you can indeed use the middleware for this.
The `isMessageMatching` method receives the incoming message and the pattern to check against.


```php
public function isMessageMatching(Message $message, $test, $regexMatched)
{
    return $message->getMessage() === $test || $message->getMessage() === strrev($test);
}
```

This would match `joke` and `ekoj`.
In my opinion, this would be sufficient to implement a fuzzy search middleware :)

Let me know if this works for you.

**Comment by christophrumpel** [comment]
Oh I got that wrong. Will checkt it immediately. thx!

**Comment by christophrumpel** [comment]
Ok I am a little bit confused. I tried it like this:

```php
$botman->group(['middleware' => new FuzzyMiddleware()], function($bot) {
    $bot->hears('fuzzy', function (BotMan $bot) {
        $bot->startConversation(new FuzzyConversation());
    });
});
```

Bot now the middleware is only called one time for the first "fuzzy" message and not for all the other questions inside the conversation.

**Comment by mpociot** [comment]
Hm...thinking more about this:

Does it make sense to use the middleware that originally was used to _start_ the conversation in the first place inside the conversation?
I mean you could potentially start the conversation from multiple callbacks / sources - each using a different middleware.

I'm not quite sure on how we would want to support middleware inside conversations because right now the middleware is only used inside the `hears` method. Conversations are a bit more more independent.

**Comment by christophrumpel** [comment]
Hopefully solved with the new BotMan middlewares #338 👍
